File: Heap Sort/Heapsort.cpp
#include "Heapsort.hpp"
#include "BinaryMaxHeap.hpp"

// O(nlog(n))
std::vector<int> heapsort(std::vector<int> array) {
  BinaryMaxHeap heap;
  std::vector<int> result;
  for (int k = 0; k < array.size(); k++) heap.insert(array[k]);
  while (heap.count() > 0) result.push_back(heap.extract());
  return result;
}

std::vector<int> heapsortInplace(std::vector<int> array) {
  int index = 0;
  int count = (int) array.size();
  while (index < count) {
    BinaryMaxHeap::heapifyUp(array, index);
    index++;
  }
  
  int temp;
  while (count > 0) {
    count--;
    temp = array[0];
    array[0] = array[count];
    array[count] = temp;
    BinaryMaxHeap::heapifyDown(array, 0, count);
  }
  
  return array;
}

File: Heap Sort/heapsort.rb
require_relative 'heap'

class Array
  def heapsort
    result = []
    heap = BinaryMinHeap.new
    self.each { |el| heap.push(el) }
    result.push(heap.extract) until heap.empty?
    result
  end

  def heapsort!
    numel = self.length
    prc = Proc.new { |parent, child| child <=> parent }
    0.upto(numel - 1) { |k| BinaryMinHeap.heapify_up(self, k, numel, &prc) }

    until numel == 0
      numel -= 1
      self[0], self[numel] = self[numel], self[0]
      BinaryMinHeap.heapify_down(self, 0, numel, &prc)
    end
  end
end